composition/test/test_dlopen_composition.py.in

```diff
@@ -12,60 +12,46 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import unittest
+
 from launch import LaunchDescription
-from launch import LaunchService
 from launch.actions import ExecuteProcess
-from launch_testing import LaunchTestService
-from launch_testing.output import create_output_lines_filter
-from launch_testing.output import create_output_test_from_file
-
+from launch.actions import OpaqueFunction
 
-def test_dlopen_composition():
-    name = 'test_dlopen_composition'
-    executable = '@DLOPEN_COMPOSITION_EXECUTABLE@'
-    talker_library = '@TALKER_LIBRARY@'
-    listener_library = '@LISTENER_LIBRARY@'
-    server_library = '@SERVER_LIBRARY@'
-    client_library = '@CLIENT_LIBRARY@'
-    output_file = '@EXPECTED_OUTPUT_ALL@'
-    launch(
-        name,
-        [executable, talker_library, listener_library, server_library, client_library],
-        output_file)
+import launch_testing
+import launch_testing.asserts
+import launch_testing_ros
 
 
-def launch(name, cmd, output_file, additional_processes=None):
-    launch_test = LaunchTestService()
+def generate_test_description(ready_fn):
     launch_description = LaunchDescription()
-
-    action = launch_test.add_fixture_action(
-        launch_description, ExecuteProcess(
-            cmd=cmd,
-            name=name,
-            output='screen'
-        )
+    process_under_test = ExecuteProcess(
+        cmd=[
+            '@DLOPEN_COMPOSITION_EXECUTABLE@',
+            '@TALKER_LIBRARY@',
+            '@LISTENER_LIBRARY@',
+            '@SERVER_LIBRARY@',
+            '@CLIENT_LIBRARY@'
+        ],
+        name='test_dlopen_composition',
+        output='screen'
     )
-    rmw_implementation = '@rmw_implementation@'
-    launch_test.add_output_test(
-        launch_description, action,
-        output_test=create_output_test_from_file(output_file),
-        output_filter=create_output_lines_filter(
-            filtered_rmw_implementation=rmw_implementation
-        )
+    launch_description.add_action(process_under_test)
+    launch_description.add_action(
+        OpaqueFunction(function=lambda context: ready_fn())
     )
-    for additional_process in (additional_processes or []):
-        launch_test.add_fixture_action(
-            launch_description, ExecuteProcess(**additional_process)
-        )
-
-    launch_service = LaunchService()
-    launch_service.include_launch_description(launch_description)
-    rc = launch_test.run(launch_service)
+    return launch_description, locals()
 
-    assert rc == 0, \
-        "The launch file failed with exit code '" + str(rc) + "'. " \
-        'Maybe the client did not receive any messages?'
 
+class TestComposition(unittest.TestCase):
```

Comment:
A pattern I use in some of our tests that use the same (or very similar) test class in multiple test files is to define it once somewhere it can be imported, and then in the test_dlopen_composition.py.in (or whatever) file just do

from somewhere import TestCompositionClass

... generate_test_description(). . .
... return whatever...

TestComposition = TestCompositionClass

The loader will pick it up as though you'd written the whole class in the file.  This would not work with the way you're substituting variables, though, so I understand why you're doing it this way.  You could parametrize the launch by passing in launch arguments via the add_launchtest cmake macro but that's going to be more lines touched for a refactor.

Reply:
Yeah, it could use a refactor for better reuse, but I'll delay it if I may.

Reply:
I agree that it's best left to a follow-up PR.